Sources/NIO/ByteBuffer-extract.swift

```diff
@@ -0,0 +1,199 @@
+        self.currentIndex = 0
+    }
+
+    public init(byteBuffer: ByteBuffer,
+                separator: UInt8) {
+        
+        self.init(byteBuffer: byteBuffer, separator: separator,
+                  start: byteBuffer.readerIndex, length: byteBuffer.readableBytes)
+    }
+}
+
+extension ByteBuffer {
+    
+    static var defaultWhitespaces = [" ", "\t"].map({UInt8($0.utf8CString[0])})
+    
+    public func sliceByTrimming(whiteSpaces: [UInt8], start: Int, length: Int) -> ByteBuffer? {
+        return withVeryUnsafeBytes { pointer -> ByteBuffer? in
+            precondition(start <= self.capacity - length)
+            
+            let buffer = pointer.bindMemory(to: UInt8.self)
+                                .dropFirst(start)
+                                .prefix(length)
+
+            // Assume that all are whitespaces in the beginning
+            var startIndex = buffer.count
+            // Advance from the start until not a whiteSpace
+            for pointee in buffer.enumerated() {
+                if pointee.element == 0 || !(whiteSpaces.contains(pointee.element)) {
+                    startIndex = pointee.offset
+                    break
+                }
+            }
+            
+            let leadingTrimmedBuffer = buffer.dropFirst(startIndex)
+            
+            // Assume that all are whitespaces
+            var endIndex = leadingTrimmedBuffer.count
+            // Retreat the ending until not a whiteSpace, null is considered a whiteSpace
+            for pointee in leadingTrimmedBuffer.reversed().enumerated() {
+                if !(whiteSpaces.contains(pointee.element) || pointee.element == 0) {
+                    endIndex = buffer.count - pointee.offset
+                    break
+                }
+            }
+            // Validate, i.e. end >= start, then return null if invalid
+            return endIndex >= startIndex ?
+                        self.getSlice(at: startIndex + start, length: endIndex - startIndex) :
+                        nil
+        }
+    }
+    
+    public func sliceByTrimmingWhitespaces() -> ByteBuffer? {
+        return sliceByTrimming(whiteSpaces: ByteBuffer.defaultWhitespaces,
+                               start: 0, length: self.readableBytes)
+    }
+    
+    /// Compares the buffer to a case insensitive `ContiguousArray<UInt8>`.
+    ///
+    /// This `ContiguousArray` could be get from applying the `asUpperCaseContiguousUTF8UIntArray`
+    ///   property on the string protocol.
+    ///
+    /// **WARNING:** MAKE SURE THAT THE STRING YOU PASS IS UPPERCASE
+    ///
+    /// This function doesn't change the indices of the buffer
+    ///
+    /// - Parameter constant: The string constant in the form of contiguous array _IN UPPER CASE_.
+    /// - Returns: Whether the ByteBuffer contains **EXACTLY** this array or no, but by ignoring case.
+    public func compareReadingToCaseInsensitiveCString<T: Collection>(_ constant: T) -> Bool
```

Comment:
This should be named `compareReadableBytes<T: Collection>(to bytes: <T>) -> Bool where T.Element == UInt8`

Reply:
Note also that the doc comment is out of date.